### PR TITLE
Adding InferenceAggregator

### PR DESCRIPTION
**Pull request by akar5h**
Minor Changes to make Analyzer work with TextSplitter for long texts.

### CODE REVIEW

**Comment by akar5h** [comment]
Hi , These are some minor changes undoing #113  , making analyzer work for long texts .

**Comment by akar5h** [comment]
However, the processing time with the current text splitter and analyzer is large for long documents, need to look for ways to optimize

**Comment by lalitpagaria** [comment]
@akar5h Thanks for the PR.
I will review this on Monday.

**Comment by lalitpagaria** [comment]
@akar5h  Until we have inference aggregator, user will never get original output. Can you please let me know what is your thought here.

I thought we are doing this -

> But now I have to clarify something, According to my experience the max_length parameter / sequence length parameter in BertModel (bert like models) does not necessarily mean text length , the text length can be beyond 512 and still the input tokens be less than 512 .
> So would it not be more optimal that instead of splitting texts, we split the tokenizer output ( tokenizer.batch_encode_plus output) like input_ids , attention mask into sequences of length 512 and feed these splits to the model ?

**Comment by akar5h** [comment]
Hi , yes that was the way to go , However since we are working with Pipeline and not directly dealing with tokenizer output and models . We cannot implement the this as it is  . 

> So would it not be more optimal that instead of splitting texts, we split the tokenizer output ( tokenizer.batch_encode_plus output) like input_ids , attention mask into sequences of length 512 and feed these splits to the model ?

So there are modules in tokenizers called pretokenizer `tokenizers.pre_tokenizers `. I used the pretokenize to estimate where to split the document , but again we can just estimate the split and not exactly split at 512 tokens since pretokenize does not has wordpiece tokenization , 
I created an alternate `textsplitter` using this pretokenize logic , https://github.com/akar5h/obsei/blob/text-splitter/obsei/preprocessor/text_splitter_v2.py , can u have a look ?  

There was no difference model processing time in both this proposed way and the cureent TextSplitter

**Comment by akar5h** [comment]
However if you follow this link : https://colab.research.google.com/github/huggingface/notebooks/blob/master/examples/question_answering.ipynb#scrollTo=n9qywopnIrJH&uniqifier=1 
Scroll down to `Preprocessing the training data` , . Here you can see how huggingface tokenizer provides a way to split longer length documents as an inbuilt operation .

**Comment by lalitpagaria** [comment]
Cool thanks Akarsh, I will have a look.
Please bear with me about delay in review, as I am bit busy in personal front.

**Comment by akar5h** [comment]
np , the last part `InterfaceAggregator` is not so straightforward since we have multi_class labels . So I'll take some time myself with that

**Comment by lalitpagaria** [comment]
@akar5h I had looked into your PR. I feel https://github.com/obsei/obsei/issues/113 okay for now. If use do not want truncation they will use TextSplitter anyway. What we need is merge module or way to merge inference after analyzer make inference.

My idea is to have multiple configuration of TextSplitter: honour sentence boundary, honour passage boundary, add title to each chunk, etc etc. Which will add value to user's pro processing capabilities.

There are few use cases some of Obsei users are asking. Major one is to determine sentiment of customer call/chat conversation.

Let me know when you are available next week we can schedule a call and discuss more.

**Comment by akar5h** [comment]
Hi @lalitpagaria , I am finalizing the Interface Aggregator module, its work in progress , I took the liberty to create a PostProcessing classes just like Preprocessors for InterfaceAggregator, will post a PR by tomorrow on this.  
`ZeroShotClassification Pipleline` does not leave us with many options with completely utilizing transformer functionalities, hence aggregation is a simple logic for now . We can have multiple configurations for different aggregation strategies in the future. 

Yes, both sentence boundary and paragraph boundaries can be implemented using `Spacy` for TextSplitter. 
> My idea is to have multiple configuration of TextSplitter: honour sentence boundary, honour passage boundary, add title to each chunk, etc etc. Which will add value to user's pro processing capabilities.


Sure , let's lets discuss over call this Saturday or coming Monday/Tuesday after 6 ?  
> There are few use cases some of Obsei users are asking. Major one is to determine sentiment of customer call/chat conversation.
> Let me know when you are available next week we can schedule a call and discuss more.

**Comment by akar5h** [comment]
@lalitpagaria , Hi , can you have a look at this merge functions . The typecheck is failing because of TextPayload.meta or BasePayload,meta . 

Kindly suggest changes  if any , and how some suggestion on how to handle the type check (mypy )

File: obsei/analyzer/ner_analyzer.py [hunk 1]
```diff
@@ -77,11 +77,11 @@ def analyze_input(
         **kwargs,
     ) -> List[TextPayload]:
         analyzer_output: List[TextPayload] = []
+        source_response_list = TextSplitter().preprocess_input(
+            source_response_list, config=TextSplitterConfig(max_split_length=512)
```
Comment (lalitpagaria) [inline obsei/analyzer/ner_analyzer.py:81]: max_split_length should be configurable ideally default value should be based on model max length

File: obsei/postprocessor/interface_aggregator.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class InterfaceAggregatorConfig(BasePostprocessorConfig):
+    aggregation_method = "most_common"
+
+    def __init__(self, **data):
+        super().__init__(**data)
+
+
+class InterfaceAggregator(BasePostprocessor):
```
Comment (lalitpagaria) [inline obsei/postprocessor/interface_aggregator.py:16]: It should be `InferenceAggregator` :)

File: obsei/postprocessor/interface_aggregator.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+from obsei.postprocessor.base_postprocessor import (
+    BasePostprocessorConfig,
+    BasePostprocessor,
+    TextPayload,
+)
+from typing import Any, List, Optional
+
+
+class InterfaceAggregatorConfig(BasePostprocessorConfig):
```
Comment (lalitpagaria) [inline obsei/postprocessor/interface_aggregator.py:9]: It should be `InferenceAggregatorConfig` :)

File: obsei/postprocessor/interface_aggregator.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    aggregation_method = "most_common"
+
+    def __init__(self, **data):
```
Comment (lalitpagaria) [inline obsei/postprocessor/interface_aggregator.py:12]: We can remove this method if there is not custom initialisation in constructor

File: obsei/postprocessor/interface_aggregator.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
+from obsei.postprocessor.base_postprocessor import (
+    BasePostprocessorConfig,
+    BasePostprocessor,
+    TextPayload,
+)
+from typing import Any, List, Optional
+
+
+class InterfaceAggregatorConfig(BasePostprocessorConfig):
+    aggregation_method = "most_common"
```
Comment (lalitpagaria) [inline obsei/postprocessor/interface_aggregator.py:10]: Let us add type check -
aggregation_method: Optional[str] = "most_common"

File: obsei/postprocessor/interface_aggregator.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class InterfaceAggregator(BasePostprocessor):
+    def postprocess_input(
+        self, input_list: List[TextPayload], config: InterfaceAggregatorConfig, **kwargs
+    ) -> List[TextPayload]:
+        prev_document_id: int = -1
+        document = []
+        aggregated_payloads = []
+        for output in input_list:
```
Comment (lalitpagaria) [inline obsei/postprocessor/interface_aggregator.py:23]: Let's move this to separate method.
This method will aggregate List of TextPayload to Dict[str, TextPayload] based on given input key.
For example this signature -
```
def payload_aggregator(input_list: List[TextPayload], key: str = "document_id") -> Dict[str, List[TextPayload]]:
```

Comment (lalitpagaria) [reply obsei/postprocessor/interface_aggregator.py:23]: Also what do you feel TextSplitter adding `chunk_id` in the meta and here we sort resultant List based on chunk_id?
It would we become useful util function in this case

**Review by lalitpagaria — CHANGES_REQUESTED** [review]
@akar5h Thank you very much for this PR.
I liked your design for Aggregator.
I have few superficial comments, we can merge this PR after that.

File: obsei/postprocessor/interface_aggregator.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    def _process_scores(
+        self, strategy: str, documents: List[TextPayload]
+    ) -> TextPayload:
+        if strategy == "most_common":
+            return self._most_common(documents)
+        if strategy == "weighted_average":
+            return self._weighted_average(documents)
+        return TextPayload(processed_text="")
+
+    def _most_common(self, documents: List[TextPayload]) -> TextPayload:
```
Comment (lalitpagaria) [inline obsei/postprocessor/interface_aggregator.py:52]: It is max of score of all the chunks? in this case name `_most_common` bit miss leading. Shouldn't we rename it to something else like `max_score` or else?

File: obsei/postprocessor/interface_aggregator.py (hunk 6 above)
Comment (lalitpagaria) [reply obsei/postprocessor/interface_aggregator.py:23]: > @lalitpagaria , Hi , can you have a look at this merge functions . The typecheck is failing because of TextPayload.meta or BasePayload,meta .
> 
> Kindly suggest changes if any , and some suggestion on how to handle the type check (mypy )

I think soon we need to standarize meta as well. But for time being let's do hackish way (please add `#TODO Properly fix this` comment there)

Before loop you can do this -
```
meta = vars(output.meta)
```

File: obsei/postprocessor/interface_aggregator.py (hunk 7 above)
Comment (akar5h) [reply obsei/postprocessor/interface_aggregator.py:52]: It isn't the maximum score of all chunks, it is the label that appeared the most number of times among n splits. Hence the name `most_common` . 
Should I keep this `most_common` strategy or add a `max_score` strat that chooses the label with the max score across n splits.

**Comment by akar5h** [comment]
The sugested changes : 
- [ ] InferenceAggregator name change 😅 
- [ ] Chunk Id, Total Chunks in splitter
- [ ] remove this method if there is not custom initialisation in constructor
- [ ] add type check 
- [ ] payload_aggregator

Todos / Feedback required :  @lalitpagaria pls comment

- [ ] finalize among `most_common` or `max_value`
- [ ] Standardise Meta or other hack for meta ( vars(meta) not working)

**Comment by lalitpagaria** [comment]
@akar5h I am doing few changes to your PR. I have some uncommitted changes. I will push them tomorrow.

**Comment by lalitpagaria** [comment]
@akar5h I have made some refactoring and added them to your PR.
Please see these make sense.

@shahrukhx01 Would you like to take a look into changes as well?

**Comment by shahrukhx01** [comment]
LGTM

**Comment by akar5h** [comment]
the refactor makes sense. LGTM.

**Review by lalitpagaria — APPROVED** [review]
LGTM!
